lib/srv/app/cloud.go

```diff
@@ -133,22 +149,27 @@ func (c *cloud) GetAWSSigninURL(ctx context.Context, req AWSSigninRequest) (*AWS
 		return nil, trace.Wrap(err)
 	}
 
-	federationURL := getFederationURL(req.TargetURL)
-	signinToken, err := c.getAWSSigninToken(ctx, &req, federationURL)
+	if req.federationURL == "" {
+		req.federationURL = getFederationURL(req.App.GetURI())
+	}
+	signinToken, err := c.getAWSSigninToken(ctx, &req)
 	if err != nil {
+		c.emitAudit(ctx, req, err)
 		return nil, trace.Wrap(err)
 	}
 
-	signinURL, err := url.Parse(federationURL)
+	signinURL, err := url.Parse(req.federationURL)
 	if err != nil {
+		c.emitAudit(ctx, req, err)
 		return nil, trace.Wrap(err)
 	}
 
+	c.emitAudit(ctx, req, nil)
```

Comment:
nit: Emitting audit event GetAWSSigninURL is not totally verbose flow. 

Why not put the audit emit audit event from handler 
?